File: backend/internal/system/system.go
// Package system reports host health for the admin panel (read-only).
package system

import (
	"os"
	"os/exec"
	"runtime"
	"strings"
	"time"

	"github.com/AbolfazlTafakori/anjam/backend/internal/config"
)

var started = time.Now()

type Memory struct {
	Total int64 `json:"total"`
	Free  int64 `json:"free"`
	RSS   int64 `json:"rss"`
}
type Disk struct {
	Free  int64 `json:"free"`
	Total int64 `json:"total"`
}
type Report struct {
	Version       string    `json:"version"`
	Node          string    `json:"node"` // runtime label; the panel prints it next to the version
	Platform      string    `json:"platform"`
	Hostname      string    `json:"hostname"`
	Uptime        int64     `json:"uptime"`
	Load          []float64 `json:"load"`
	CPUs          int       `json:"cpus"`
	Memory        Memory    `json:"memory"`
	Disk          *Disk     `json:"disk"`
	DBBytes       int64     `json:"dbBytes"`
	PublicURL     string    `json:"publicUrl"`
	AdminPath     string    `json:"adminPath"`
	CertExpiresAt *int64    `json:"certExpiresAt"`
	DataDir       string    `json:"dataDir"`
}

func Info(cfg *config.Config, dbBytes int64) Report {
	host, _ := os.Hostname()
	var ms runtime.MemStats
	runtime.ReadMemStats(&ms)
	r := Report{
		Version: cfg.Version, Node: "Go " + runtime.Version(), Platform: runtime.GOOS + " " + runtime.GOARCH, Hostname: host,
		Uptime: int64(time.Since(started).Seconds()), Load: loadAvg(), CPUs: runtime.NumCPU(),
		Memory: Memory{Total: memTotal(), Free: memFree(), RSS: int64(ms.Sys)}, Disk: diskFree(cfg.DataDir), DBBytes: dbBytes,
		PublicURL: cfg.PublicURL, AdminPath: cfg.AdminPath, DataDir: cfg.DataDir, CertExpiresAt: certExpiry(cfg.PublicURL),
	}
	return r
}

func certExpiry(publicURL string) *int64 {
	host := strings.TrimPrefix(strings.TrimPrefix(publicURL, "https://"), "http://")
	if i := strings.IndexByte(host, '/'); i >= 0 {
		host = host[:i]
	}
	pem := "/etc/letsencrypt/live/" + host + "/cert.pem"
	if host == "" {
		return nil
	}
	if _, err := os.Stat(pem); err != nil {
		return nil
	}
	out, err := exec.Command("openssl", "x509", "-enddate", "-noout", "-in", pem).Output()
	if err != nil {
		return nil
	}
	s := strings.TrimSpace(strings.TrimPrefix(string(out), "notAfter="))
	t, err := time.Parse("Jan 2 15:04:05 2006 MST", s)
	if err != nil {
		return nil
	}
	v := t.UnixMilli()
	return &v
}

File: server/src/system.js
'use strict';
/* Server health for the panel's overview: host, runtime, memory, disk, certificate. Read-only. */
const os = require('os');
const fs = require('fs');
const { execFileSync } = require('child_process');
const config = require('./config');

function diskFree(dir) {
  try { const s = fs.statfsSync(dir); return { free: s.bavail * s.bsize, total: s.blocks * s.bsize }; } catch { return null; }
}
function certExpiry() {
  const host = config.publicUrl.replace(/^https?:\/\//, '').split('/')[0];
  if (!host) return null;
  const pem = `/etc/letsencrypt/live/${host}/cert.pem`;
  if (!fs.existsSync(pem)) return null;
  try {
    const out = execFileSync('openssl', ['x509', '-enddate', '-noout', '-in', pem], { encoding: 'utf8', timeout: 3000 });
    const d = new Date(out.split('=')[1].trim()); return isNaN(d) ? null : d.getTime();
  } catch { return null; }
}
function info() {
  return {
    version: config.version, node: process.version, platform: `${os.type()} ${os.release()} ${os.arch()}`, hostname: os.hostname(),
    uptime: Math.round(process.uptime()), osUptime: os.uptime(), load: os.loadavg().map((x) => +x.toFixed(2)), cpus: os.cpus().length,
    memory: { total: os.totalmem(), free: os.freemem(), rss: process.memoryUsage().rss },
    disk: diskFree(config.dataDir), dbBytes: fs.existsSync(config.dbFile) ? fs.statSync(config.dbFile).size : 0,
    publicUrl: config.publicUrl, adminPath: config.adminPath, certExpiresAt: certExpiry(), dataDir: config.dataDir,
  };
}
module.exports = { info };
